playbooks/operationalizing/how-full-is-my-cluster.adoc

```diff
@@ -0,0 +1,150 @@
+
+This will produce an output of the form (only the relevant fragments are reported): +
+```
+Allocatable: 
+ alpha.kubernetes.io/nvidia-gpu:  0 
+ cpu:         2 
+ memory:        12304368Ki 
+ pods:          20 
+... 
+Allocated resources: 
+  (Total limits may be over 100 percent, i.e., overcommitted. 
+  CPU Requests  CPU Limits  Memory Requests Memory Limits +
+  ------------  ----------  --------------- ------------- 
+  300m (15%)  400m (20%)  2660Mi (22%)  2248Mi (18%)
+```
+In this particular example, we have a node with 12GB of memory, of which 2.6GB are booked by requests coming from the 20 pods that are deployed.  +
+Notice that this figure (22%) has nothing to do with the actual memory used on the node. This value represents the memory that has been reserved. The same applies to the CPU. +
+Also, if you don’t specify what amount of resources you container needs, OpenShift assumes zero. This implies that if you don’t specify the requests for your containers, you can easily run out of resources while OpenShift will still think that all your nodes have capacity available.
+
+*Best practice:* specify the request for all of your containers. You can mandate that requests for all your containers be specified by setting a min request in the limit range object associated to your projects, here is an example:
+
+```
+apiVersion: "v1" 
+kind: "LimitRange" 
+metadata: 
+  name: "core-resource-limits"  
+spec: 
+  limits: 
+    - type: "Container" 
+      min: 
+        cpu: "1m"  
+        memory: "1Mi"
+``` 
+In the above example very small minimums are specified that will not affect the value of the requests for containers, but will make it mandatory to specify it.
+
+It is also worth noting that in version 3.5, a link:https://docs.openshift.com/container-platform/latest/dev_guide/compute_resources.html#opaque-integer-resources[generalized framework for tracking additional node resources] has been introduced. These resources are called opaque integer resources and the main use case is to be able to track link:https://blog.openshift.com/use-gpus-openshift-kubernetes/[GPUs’ resources].
+
+Every time OpenShift places a pod, it is solving an instance of a multidimensional link:https://en.wikipedia.org/wiki/Knapsack_problem[knapsack problem]. The knapsack problem is a classical problem in algorithm theory, where you have to place N stones of different sizes in M backpacks of different capacity. The point is to find optimal allocation. In the case of OpenShift we have a multidimensional knapsack problem because there are more that one characteristic to consider: CPU, memory and as we have seen opaque integer resources. +
+The knapsack problem is a link:https://en.wikipedia.org/wiki/NP-completeness[np-complete] problem, which means that time needed to solve it scales exponentially with n (pods) and m (nodes). For this reason when n and m are big enough, a human cannot do a better job than a machine at solving this problem.
```

Comment:
either add a line break to create a paragraph, or remove the line break.

Reply:
done